packages/react/src/hooks/useNotifications.ts

```diff
@@ -92,26 +150,47 @@ export const useNotifications = (props?: UseNotificationsProps): UseNotification
   const length = data?.length;
   const after = length ? data[length - 1].id : undefined;
 
-  useWebSocketEvent({
-    event: 'notifications.unread_count_changed',
-    eventHandler: () => {
-      void refetch();
-    },
-  });
+  useEffect(() => {
+    const listener = ({
+      data: updatedData,
+    }: {
+      data: { notifications: Notification[]; hasMore: boolean; filter: NotificationFilter };
+    }) => {
+      if (!updatedData) return;
 
-  useWebSocketEvent({
-    event: 'notifications.unseen_count_changed',
-    eventHandler: () => {
-      void refetch();
-    },
-  });
+      const currentFilter = getCurrentFilter();
+      const filterMatches = isSameFilter(currentFilter, updatedData.filter);
+
+      // Always merge - mergeNotifications handles filtering correctly
+      setData((prev) => mergeNotifications(prev, updatedData.notifications, currentFilter, filterMatches));
+
+      // Only update hasMore when filter matches exactly (preserves pagination)
+      if (filterMatches) {
+        setHasMore(updatedData.hasMore);
+      }
+    };
+
+    const cleanup = on('notifications.list.updated', listener);
+
+    return () => {
+      cleanup();
+    };
+  }, [getCurrentFilter, on]);
 
   useWebSocketEvent({
     event: 'notifications.notification_received',
     eventHandler: ({ result: notification }) => {
       const currentFilter = getCurrentFilter();
       const matches = checkNotificationMatchesFilter(notification, currentFilter);
-      if (matches) void refetch();
+      if (matches) {
+        setData((prev) => {
+          // Prevent duplicates - mergeNotifications will handle this when notifications.list.updated fires
+          if (prev?.some((n) => n.id === notification.id)) {
+            return prev;
+          }
+          return prev ? [notification, ...prev] : [notification];
+        });
+      }
```

Comment:
I am not sure if you saw this comment before, but I was thinking about doing it [like this](https://novu.slack.com/archives/C06L08A4USH/p1753915931080109?thread_ts=1753033671.138029&cid=C06L08A4USH)